### DISCUSSION TITLE
Variable MIDIAddress?

### DISCUSSION THREAD

**Discussion by fitzgreyve** [tttapa/Control-Surface#582] [General]
Hi,

Is there an elegant method to create a MIDIAddress (for use in sendNoteOn, sendNoteoff etc) from variables dynamically _at runtime_ ?

If not, what is the best method to generate MIDI events with fully variable values  (e.g. MIDI channel value , note value).

In summary, I am generating the required MIDI values from combinations of FilteredAnalog.

Many thanks for any help on this.

**Comment by tttapa** [comment]
I'm not entirely sure what you mean, there are no restrictions on how you can create a MIDIAddress, e.g.:
```cpp
    for (uint8_t note : {12, 24, 36}) {
        for (Channel ch : {CHANNEL_1, CHANNEL_2}) {
            MIDIAddress addr = {note, ch};
            midi.sendNoteOn(addr, 127);
        }
    }
```

**Reply by fitzgreyve** [reply]
Many thanks for the fast response.
I've got this working fine for a variable note value and a constant MIDI channel e.g.
```c++
    noteNo=cvs[0].getValue()+panelPotentiometers[0].getValue()-64;    // or any other calculated value
    MIDIAddress addr = {noteNo, CHANNEL_1};
     serialMIDI.sendNoteOn(addr, 127);
```

but if I try and amend this for a variable MIDI channel value:
```c++
    noteNo=cvs[0].getValue()+panelPotentiometers[0].getValue()-64;    // or any other calculated value
    uint8_t channelNo=1;
    ++channelNo;                                                      // or any other calculated value 
    MIDIAddress addr = {noteNo, channelNo};   
    serialMIDI.sendNoteOn(addr, 127);
```
results in error message:  "could not convert '{noteNo, channelNo}' from '<brace-enclosed initializer list>' to 'CS::MIDIAddress'"

I know I'm missing something very fundemental here in the way the control surface library and c++ works - can you "point me in the right direction"?

**Reply by tttapa** [reply]
A channel is not an integer, it's a custom type: https://tttapa.github.io/Control-Surface-doc/Doxygen/d2/d48/classChannel.html

The reason is that the raw channel number sent over MIDI is zero-based (i.e. 0 - 15), while the MIDI standard specifies that the channels should be referred to as channel 1 - 16.

You can convert a zero-based integer to a channel by using the explicit `Channel` constructor:
```cpp
    noteNo=cvs[0].getValue()+panelPotentiometers[0].getValue()-64;    // or any other calculated value
    uint8_t channelNo=0; // channel 1
    ++channelNo;                                                      // or any other calculated value 
    Channel channel {channelNo};
    MIDIAddress addr = {noteNo, channel};   
    serialMIDI.sendNoteOn(addr, 127);
```
Or in a single line: `MIDIAddress addr = {noteNo, Channel{channelNo}};`

**Reply by fitzgreyve** [reply]
THanks, I had tried to use Channel,  but I used  incorrect syntaxt. It now works. THanks again for your help.
